core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,493 @@
+
+	return strings.Join(clauses, " "), clauseArgs, nil
+}
+
+func (v *pgDSLParser) selectClause(props *dslQueryProps) string {
+	fields := []string{fmt.Sprintf("%s.*", evmLogsTableName)}
+
+	// if order by has sequence or if cursor limit is used, add sequence to fields
+	if props.hasCursorLimit() || props.hasSequenceSort() {
+		fields = append(fields, fmt.Sprintf("%s AS %s", sequenceField, sequenceFieldAlias))
+	}
+
+	return fmt.Sprintf("SELECT %s FROM %s", strings.Join(fields, ", "), evmLogsTableName)
+}
+
+func (v *pgDSLParser) whereClause(props *dslQueryProps) (string, map[string]any) {
+	segment := fmt.Sprintf(
+		"WHERE evm_chain_id = :%s",
+		v.args.Add("evm_chain_id", props.ChainID.Uint64()),
+	)
+
+	if len(props.Filter.Expressions) > 0 {
+		segment = fmt.Sprintf("%s AND %s", segment, v.combineExpressions(props.Filter.Expressions, query.AND))
+	}
+
+	if props.hasCursorLimit() {
+		var op string
+		switch *props.Limiter.Limit.CursorDirection {
+		case query.Following:
+			op = ">"
+		case query.Previous:
+			op = "<"
+		default:
+			op = "<"
+		}
+
+		segment = fmt.Sprintf(
+			"%s AND %s %s :%s",
+			segment,
+			sequenceFieldAlias,
+			op,
+			v.args.Add(sequenceFieldAlias, *props.Limiter.Limit.Cursor),
+		)
+	}
+
+	return segment, v.args.Values()
+}
+
+func (v *pgDSLParser) orderClause(props *dslQueryProps) string {
+	sorting := props.Limiter.SortBy
+
+	if props.hasCursorLimit() && !props.hasSequenceSort() {
+		sorting = append(sorting, query.NewSortBySequence(query.Asc))
+	}
+
+	if len(sorting) == 0 {
+		return ""
+	}
+
+	sort := make([]string, len(sorting))
+
+	for idx, sorted := range sorting {
+		var name string
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			name = sequenceFieldAlias
+		case query.SortByTimestamp:
+			name = timestampFieldName
+		default:
+			continue
+		}
+
+		sort[idx] = fmt.Sprintf("%s %s", name, orderToString(sorted.GetDirection()))
+	}
+
+	return fmt.Sprintf("ORDER BY (%s)", strings.Join(sort, ", "))
+}
+
+func (v *pgDSLParser) limitClause(props *dslQueryProps) string {
+	if !props.hasCursorLimit() && props.Limiter.Limit.Count == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf("LIMIT %d", props.Limiter.Limit.Count)
+}
+
+func (v *pgDSLParser) getLastExpression() string {
+	exp := v.expression
+	v.expression = ""
+
+	return exp
+}
+
+func (v *pgDSLParser) combineExpressions(expressions []query.Expression, op query.BoolOperator) string {
+	clauses := []string{}
+	grouped := len(expressions) > 1
+
+	for _, exp := range expressions {
+		if exp.IsPrimitive() {
+			exp.Primitive.Accept(v)
+
+			clauses = append(clauses, v.getLastExpression())
+		} else {
+			clauses = append(clauses, v.combineExpressions(exp.BoolExpression.Expressions, exp.BoolExpression.BoolOperator))
+		}
+	}
+
+	output := strings.Join(clauses, fmt.Sprintf(" %s ", op.String()))
+
+	if grouped {
+		output = fmt.Sprintf("(%s)", output)
+	}
+
+	return output
+}
+
+type namedArgs struct {
```

Comment:
Can't we reuse what log poller already has?

```
 type queryArgs struct {
	args map[string]interface{}
	err  []error
}
```

Reply:
I didn't really want to modify the existing query args, but I'll include a commit just for this and we can see how it goes.